session_db/session.py

```diff
@@ -0,0 +1,126 @@
+# Copyright (c) Odoo SA 2017
+# @author Nicolas Seinlet
+# Copyright (c) ACSONE SA 2022
+# @author Stéphane Bidoul
+import json
+import logging
+import os
+
+import psycopg2
+
+import odoo
+from odoo import http
+from odoo.tools._vendor import sessions
+from odoo.tools.func import lazy_property
+
+_logger = logging.getLogger(__name__)
+
+
+def with_cursor(func):
+    def wrapper(self, *args, **kwargs):
+        tries = 0
+        while True:
+            tries += 1
+            try:
+                return func(self, *args, **kwargs)
+            except psycopg2.InterfaceError as e:
+                _logger.info("Session in DB connection Retry %s/5" % tries)
+                if tries > 4:
+                    raise e
+                self._open_connection()
+
+    return wrapper
+
+
+class PGSessionStore(sessions.SessionStore):
+    def __init__(self, uri, session_class=None):
+        super().__init__(session_class)
+        self._uri = uri
+        self._cr = None
+        # FIXME This class is NOT thread-safe. Only use in worker mode
+        if odoo.tools.config["workers"] == 0:
+            raise ValueError("session_db requires multiple workers")
+        self._open_connection()
+        self._setup_db()
+
+    def __del__(self):
+        if self._cr is not None:
+            self._cr.close()
+
+    def _open_connection(self):
+        cnx = odoo.sql_db.db_connect(self._uri, allow_uri=True)
+        self._cr = cnx.cursor()
+        self._cr._cnx.autocommit = True
+
+    @with_cursor
+    def _setup_db(self):
+        self._cr.execute(
+            """
+            CREATE TABLE IF NOT EXISTS http_sessions (
+                sid varchar PRIMARY KEY,
+                write_date timestamp without time zone NOT NULL,
+                payload text NOT NULL
+            )
+        """
+        )
+
+    @with_cursor
+    def save(self, session):
+        payload = json.dumps(dict(session))
+        self._cr.execute(
+            """
+                INSERT INTO http_sessions(sid, write_date, payload)
+                    VALUES (%(sid)s, now() at time zone 'UTC', %(payload)s)
+                ON CONFLICT (sid)
+                DO UPDATE SET payload = %(payload)s,
+                              write_date = now() at time zone 'UTC'
+            """,
+            dict(sid=session.sid, payload=payload),
+        )
+
+    @with_cursor
+    def delete(self, session):
+        self._cr.execute("DELETE FROM http_sessions WHERE sid=%s", [session.sid])
+
+    @with_cursor
+    def get(self, sid):
+        self._cr.execute(
+            "UPDATE http_sessions "
+            "SET write_date = now() at time zone 'UTC' "
+            "WHERE sid=%s",
+            [sid],
+        )
```

Comment:
Updating the write_date when reading is likely not necessary. If I read the code correctly the FilesystemSessionStore does not do that.

Reply:
The point of doing this is to have the session not expire after 8 days.

Reply:
Yes, but the FilesystemSessionStore does not do that right? Any specific reason to do it differently here?